core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -143,24 +143,26 @@ private JoinNode getSizeBasedJoin(JoinNode joinNode, Context context)
     {
         DataSize joinMaxBroadcastTableSize = getJoinMaxBroadcastTableSize(context.getSession());
 
-        boolean isRightSideSmall = getSourceTablesSizeInBytes(joinNode.getRight(), context) <= joinMaxBroadcastTableSize.toBytes();
+        double rightSourceSize = getSourceTablesSizeInBytes(joinNode.getRight(), context);
+        boolean isRightSideSmall = rightSourceSize <= joinMaxBroadcastTableSize.toBytes();
         if (isRightSideSmall && !mustPartition(joinNode)) {
             // choose right join side with small source tables as replicated build side
             return joinNode.withDistributionType(REPLICATED);
         }
 
-        boolean isLeftSideSmall = getSourceTablesSizeInBytes(joinNode.getLeft(), context) <= joinMaxBroadcastTableSize.toBytes();
+        double leftSourceSize = getSourceTablesSizeInBytes(joinNode.getLeft(), context);
+        boolean isLeftSideSmall = leftSourceSize <= joinMaxBroadcastTableSize.toBytes();
         if (isLeftSideSmall && !mustPartition(joinNode.flipChildren())) {
             // choose join left side with small source tables as replicated build side
             return joinNode.flipChildren().withDistributionType(REPLICATED);
         }
 
-        if (isRightSideSmall) {
+        if ((isRightSideSmall || rightSourceSize < leftSourceSize * 4) && !mustReplicate(joinNode, context)) {
             // right side is small enough, but must be partitioned
             return joinNode.withDistributionType(PARTITIONED);
         }
 
-        if (isLeftSideSmall) {
+        if ((isLeftSideSmall || leftSourceSize < rightSourceSize * 4) && !mustReplicate(joinNode, context)) {
```

Comment:
why 4?

replace `mustReplicate` with `mustPartition`. Use `flipChildren`

Reply:
I've re-arranged the code a bit, we need to use `mustReplicate` to avoid using PARTITIONED type on a cross join.